src/lib.rs

```diff
@@ -1113,21 +1083,67 @@ pub mod invariant {
         }
 
         #[ink(message)]
-        fn get_liquidity_ticks_amount(&self, pool_key: PoolKey) -> u32 {
+        fn get_liquidity_ticks_amount(
+            &self,
+            pool_key: PoolKey,
+            lower_tick: i32,
+            upper_tick: i32,
+        ) -> Result<u32, InvariantError> {
             let tick_spacing = pool_key.fee_tier.tick_spacing;
+            if tick_spacing == 0 {
+                return Err(InvariantError::InvalidTickSpacing);
+            };
+
+            if lower_tick % (tick_spacing as i32) != 0 || upper_tick % (tick_spacing as i32) != 0 {
+                return Err(InvariantError::InvalidTickIndex);
+            }
 
             let max_tick = get_max_tick(tick_spacing);
-            let (chunk_limit, _) = tick_to_position(max_tick, tick_spacing);
+            let min_tick = get_min_tick(tick_spacing);
+
+            if lower_tick < min_tick || upper_tick > max_tick {
+                return Err(InvariantError::InvalidTickIndex);
+            };
+
+            let (min_chunk_index, min_bit) = tick_to_position(lower_tick, tick_spacing);
+            let (max_chunk_index, max_bit) = tick_to_position(upper_tick, tick_spacing);
+
+            let active_bits_in_range = |chunk, min_bit, max_bit| {
+                let mut amount = 0;
+                for bit in min_bit..=max_bit {
```

Comment:
This does not look like an optimal function because it uses loops underneath. It is best to use the `count_ones` function.

The `count_ones()` leverages hardware instructions referred to as the "population count" instruction. POPCNT - Return the Count of Number of Bits Set to 1.

For example something similar to this:
```Rust
let active_bits_in_range = |chunk, min_bit, max_bit| -> u32 {
    ((chunk >> min_bit) & ((1 << (max_bit - min_bit + 1)) - 1)).count_ones()
};
```